### PR TITLE
Add Several Hold Effects

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
Adds:
- Throat Spray
throat_spray
- Blunder Policy
blunder_policy
- Room Service
room_service
- Terrain Seeds
seed
- Heavy Duty Boots (no gif)
- Red Card
red_card
- Eject Button (no gif)
- Eject Pack
eject-pack

### CODE REVIEW

File: src/battle_util.c [hunk 1]
```diff
@@ -5172,6 +5172,33 @@ static u8 RandomStatRaiseBerry(u32 battlerId, u32 itemId)
...
 }
 
+u8 TryHandleSeed(u8 battler, u32 terrainFlag, u8 statId, u16 itemId, bool32 execute)
+{
+    if (gFieldStatuses & terrainFlag && gBattleMons[battler].statStages[statId] < MAX_STAT_STAGE)
+    {
+        PREPARE_STAT_BUFFER(gBattleTextBuff1, statId);
+        PREPARE_STRING_BUFFER(gBattleTextBuff2, STRINGID_STATROSE);
+
+        gLastUsedItem = itemId; // for surge abilities
```
Comment (ultima-soul) [inline src/battle_util.c:5182]:
```suggestion
        gLastUsedItem = itemId; // For surge abilities
```

File: src/battle_util.c [hunk 2]
```diff
@@ -5398,6 +5425,54 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
...
+                case HOLD_EFFECT_PARAM_GRASSY_TERRAIN:
+                    effect = TryHandleSeed(battlerId, STATUS_FIELD_GRASSY_TERRAIN, STAT_DEF, gLastUsedItem, TRUE);
+                    break;
+                case HOLD_EFFECT_PARAM_MISTY_TERRAIN:
+                    effect = TryHandleSeed(battlerId, STATUS_FIELD_MISTY_TERRAIN, STAT_SPDEF, gLastUsedItem, TRUE);
+                    break;
+                case HOLD_EFFECT_PARAM_PSYCHIC_TERRAIN:
+                    effect = TryHandleSeed(battlerId, STATUS_FIELD_PSYCHIC_TERRAIN, STAT_SPDEF, gLastUsedItem, TRUE);
+                    break;
+
```
Comment (ultima-soul) [inline src/battle_util.c:5456]:
```suggestion
```

File: src/battle_util.c [hunk 3]
```diff
@@ -5398,6 +5425,54 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
...
+                    effect = ITEM_STATS_CHANGE;
+                    if (moveTurn)
+					{
+						BattleScriptPushCursor();
+						gBattlescriptCurrInstr = BattleScript_EjectPackActivate_Ret;
+					}
+					else
+                    {
+						BattleScriptExecute(BattleScript_EjectPackActivate_End2);
+                    }
```
Comment (ultima-soul) [inline src/battle_util.c:5473]:
```suggestion
                    {
                        BattleScriptPushCursor();
                        gBattlescriptCurrInstr = BattleScript_EjectPackActivate_Ret;
                    }
                    else
                    {
                        BattleScriptExecute(BattleScript_EjectPackActivate_End2);
                    }
```

File: src/battle_util.c [hunk 4]
```diff
@@ -5837,44 +5912,69 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
         }
         break;
     case ITEMEFFECT_KINGSROCK_SHELLBELL:
-        if (gBattleMoveDamage)
+        switch (atkHoldEffect)
         {
-            switch (atkHoldEffect)
+        case HOLD_EFFECT_FLINCH:
+            if (gBattleMoveDamage != 0  // need to have done damage
```
Comment (ultima-soul) [inline src/battle_util.c:5918]:
```suggestion
            if (gBattleMoveDamage != 0  // Need to have done damage
```

File: src/battle_util.c [hunk 5]
```diff
@@ -5837,44 +5912,69 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
...
-                }
-                break;
+                gBattleScripting.moveEffect = MOVE_EFFECT_FLINCH;
+                BattleScriptPushCursor();
+                SetMoveEffect(FALSE, 0);
+                BattleScriptPop();
             }
+            break;
+        case HOLD_EFFECT_SHELL_BELL:
+            if (gBattleMoveDamage != 0  // need to have done damage
```
Comment (ultima-soul) [inline src/battle_util.c:5932]:
```suggestion
            if (gBattleMoveDamage != 0  // Need to have done damage
```

File: src/battle_util.c [hunk 6]
```diff
@@ -5837,44 +5912,69 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
...
+                gBattleMoveDamage = (gSpecialStatuses[gBattlerTarget].dmg / atkHoldEffectParam) * -1;
+                if (gBattleMoveDamage == 0)
+                    gBattleMoveDamage = -1;
+                gSpecialStatuses[gBattlerTarget].dmg = 0;
+                BattleScriptPushCursor();
+                gBattlescriptCurrInstr = BattleScript_ItemHealHP_Ret;
+                effect++;
+            }
+            break;
+        case HOLD_EFFECT_THROAT_SPRAY:  // doesn't need to be a damaging move
```
Comment (ultima-soul) [inline src/battle_util.c:5952]:
```suggestion
        case HOLD_EFFECT_THROAT_SPRAY:  // Doesn't need to be a damaging move
```

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Notes from testing:
```
Throat Spray:
    - Should raise Sp. Atk of holder, not opponent
    - If target is asleep and Sing fails due to already being asleep, shouldn't activate
    - If target has Soundproof and fails, shouldn't activate
    - If target fainting ends the battle (last mon of the opponent), shouldn't activate
    - Should activate after both stat drops in double battle when something like Growl used

Room Service:
    - When switched in and holding Room Service and Trick Room is in effect, item use animation should play and stat lowering sound should be used (unless has Contrary)

Terrain Seeds:
    - Switching in Contrary mon with Seed that can activate should have stat decrease sound to match stat decrease
    - Contrary mon with +6 Defense/Sp. Def (depending on the Seed) should have Seed activate
    - Contrary mon with -6 Defense/Sp. Def (depending on the Seed) shouldn't have Seed activate

Heavy-Duty Boots:
    - Grounded Poison-types when switched in should still absorb Toxic Spikes

Red Card:
    - If Wild mon holds Red Card, it should activate and make player flee if hit
    - If Wild mon holds Red Card in double wild battle with two enemies active, it should activate and make player switch (if there is a third mon in the party minimum)
        - If single enemy active, instead makes player flee like single battle
    - In double battle, if hitting with attack that hits multiple mons all holding Red Card, only fastest Red Card holder activates its item
    - In double battle, if hitting with attack that hits multiple mons all holding Red Card, if at least one mon isn't hit due to Protect or evading the attack, no Red Card activation takes place when it should activate
    - If attacker that has used Ingrain and is rooted, attacks holder of Red Card, Red Card should get used up, but fails to make attacker switch out (just like how it works with attacker having Suction Cups)
    - If attacker holds Shell Bell, it should not activate if holder of Red Card is attacked and activates its Red Card (Like holding Life Orb)

Eject Button:
    - Being trapped by trapping moves like Fire Spin should not prevent Eject Button from activating
    - In double battle, if hitting with attack that hits multiple mons all holding Red Card, if at least one mon isn't hit due to Protect or evading the attack, no Eject Button activation takes place when it should activate

Eject Pack:
    - If Parting Shot is used against holder of Eject Pack and user is able to switch out, Eject Pack doesn't activate
    - If U-turn or similar is used against holder of Eject Pack that has Weak Armor ability and user is able to switch out, Eject Pack doesn't activate
    - Stockpile stat changes that get lowered by Swallow/Spit Up should activate Eject Pack
```

**Comment by ghoulslash** [comment]
Ready for re-review.

The recent commits also take care of:
- fixes shell bell, throat spray, etc, activation timing relative to multi-hit moves
- handles contrary checks with the function `CompareStat`

Eject Pack + swallow/spit up interaction is not handled because these are not updated to Gen 5 behavior where the stat drop is handled as an animation. This should be a separate PR (probably after this PR is merged to handle eject pack)

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Some issues that remain:
```
Throat Spray
    - Should activate after hitting at least one target (and other target it can miss against or be protected against)

Room Service
    - When switched in and holding Room Service and Trick Room is in effect and has Contrary, stat increase and its animation should happen

Terrain Seeds
    - Switching in Contrary mon with Seed that can activate, stat decrease and its animation should happen

Red Card
    - If Wild mon holds Red Card in double wild battle with two enemies active:
        - If player has at least a third Pokemon to switch to and attacking player's Pokemon is alive at the time Red Card is activated, but attacker's partner fainted before it in the same turn:
            - Should switch attacking Pokemon to one of your available party Pokemon
            - Then choose Pokemon to replace fainted partner Pokemon at end of turn
        -If no available party Pokemon to switch to (whether attacker's partner faints in the same turn or is alive):
            - Red Card shouldn't activate
    - If Wild mon holds Red Card in double wild battle with one enemy active:
        - If player has at least a third Pokemon to switch to and both of player's Pokemon are alive at the time Red Card is activated:
            - Should switch attacking Pokemon to one of your available party Pokemon
        - If player has at least a third Pokemon to switch to and attacking player's Pokemon is alive at the time Red Card is activated, but attacker's partner fainted before it in the same turn:
            - Should switch attacking Pokemon to one of your available party Pokemon
            - Then choose Pokemon to replace fainted partner Pokemon at end of turn
        -If no available party Pokemon to switch to (whether attacker's partner faints in the same turn or is alive):
            - Red Card shouldn't activate
    - If holder of Red Card (whether a trainer's mon or a wild mon) is hit and attacker's trainer has no mons to switch to, it shouldn't activate

Eject Pack
    - If Parting Shot is used against holder of Eject Pack and user of move is able to switch out, Eject Pack shouldn't activate
    - If U-turn or similar is used against holder of Eject Pack that has Weak Armor ability and user is able to switch out, Eject Pack doesn't activate
    ```
    
For full notes on how Red Card works from official game testing take a look at: https://gist.github.com/ultima-soul/3fa958515961081889ed9af6c3385671

File: data/battle_scripts_1.s [hunk 7]
```diff
@@ -1842,6 +1843,35 @@ BattleScript_EffectMagnetRise:
...
+	attackstring
+	ppreduce
+	setroom
+	attackanimation
+	waitanimation
+	printfromtable gRoomsStringIds
+	waitmessage 0x40
+	savetarget
+	setbyte gBattlerTarget, 0
+RoomServiceLoop:
```
[the remark this answers is not in the document]
Comment (ultima-soul) [reply data/battle_scripts_1.s]: Also unresolved this because `TerrainSeedLoop` and `TerrainSeedLoop_NextBattler` are still outstanding for this issue.

**Comment by ghoulslash** [comment]
ready for re-review

**Review by ultima-soul — CHANGES_REQUESTED** [review]
All issues are fixed. Just resolve the suggested changes and also the remaining suggested changes from the previous review.

**Comment by ghoulslash** [comment]
ready for review

**Review by ultima-soul — APPROVED** [review]

**Pull request merged**